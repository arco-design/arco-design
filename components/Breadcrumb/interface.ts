import React, { CSSProperties, ReactNode, HTMLAttributes } from 'react';
import { DropdownProps } from '../Dropdown';

export interface RouteProps {
  path: string;
  breadcrumbName: string;
  children?: Array<{ path: string; breadcrumbName: string }>;
}

/**
 * @title Breadcrumb
 */
export interface BreadcrumbProps {
  style?: CSSProperties;
  children?: ReactNode;
  className?: string | string[];
  /**
   * @zh 指定分割符
   * @en Custom separator
   * @defaultValue <IconObliqueLine />
   */
  separator?: string | ReactNode;
  /**
   * @zh 直接设置下拉菜单
   * @en Set drop-down menu
   */
  routes?: RouteProps[];
  /**
   * @zh 最多渲染的面包屑数量
   * @en Max count of `Breadcrumb.Item` to show.
   */
  maxCount?: number;
  /**
   * @zh routes 时生效，自定义渲染面包屑
   * @en Custom render function for `Breadcrumb.Item`
   */
  itemRender?: (route: RouteProps, routes: RouteProps[], paths: string[]) => ReactNode;
}

type CustomHTMLElement = Omit<HTMLAttributes<HTMLDivElement>, 'className'>;

/**
 * @title Breadcrumb.Item
 */
export interface BreadCrumbItemProps extends CustomHTMLElement {
  style?: CSSProperties;
  className?: string | string[];
  prefixCls?: string;
  /**
   * @zh 下拉菜单的内容，等同于下拉菜单组件的 droplist 属性
   * @en The dropdown menu
   */
  droplist?: DropdownProps['droplist'];
  /**
   * @zh 下拉菜单的属性 [DropdownProps](/react/components/dropdown)
   * @en The dropdown props [DropdownProps](/react/components/dropdown)
   */
  dropdownProps?: DropdownProps;
  /**
   * @zh 超链接地址
   * @en href
   * @version 2.40.0
   */
  href?: string;
  /**
   * @zh 点击回调
   * @en click callback
   * @version 2.40.0
   */
  onClick?: (e: any) => void;
  /**
   * @zh 标签名，可以是 html 标签或是组件
   * @en Configure the outermost label, which can be an html label or a component
   * @version 2.40.0
   * @defaultValue div
   */
  tagName?: string | React.FC<any> | React.ComponentClass<any>;
}
