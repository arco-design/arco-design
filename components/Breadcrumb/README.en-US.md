`````
Component / Navigation

# Breadcrumb

Used to display the location of the current page and quickly return to the history page.
`````

%%Content%%

## API

### Breadcrumb

|Property|Description|Type|DefaultValue|
|---|---|---|---|
|maxCount|Max count of `Breadcrumb.Item` to show.|number |`-`|
|separator|Custom separator|string \| ReactNode |`<IconObliqueLine />`|
|className|Additional css class|string \| string[] |`-`|
|routes|Set drop-down menu|[RouteProps](#routeprops)[] |`-`|
|style|Additional style|CSSProperties |`-`|
|itemRender|Custom render function for `Breadcrumb.Item`|(route: [RouteProps](#routeprops), routes: [RouteProps](#routeprops)[], paths: string[]) => ReactNode |`-`|

### Breadcrumb.Item

|Property|Description|Type|DefaultValue|Version|
|---|---|---|---|---|
|href|href|string |`-`|2.40.0|
|droplist|The dropdown menu|[DropdownProps](dropdown#dropdown)['droplist'] |`-`|-|
|className|Additional css class|string \| string[] |`-`|-|
|dropdownProps|The dropdown props [DropdownProps](/react/components/dropdown)|[DropdownProps](dropdown#dropdown) |`-`|-|
|style|Additional style|CSSProperties |`-`|-|
|tagName|Configure the outermost label, which can be an html label or a component|string \| React.FC&lt;any&gt; \| React.ComponentClass&lt;any&gt; |`div`|2.40.0|
|onClick|click callback|(e: any) => void |`-`|2.40.0|

### RouteProps

```js
export interface RouteProps {
  path: string;
  breadcrumbName: string;
  children?: Array<{
    path: string;
    breadcrumbName: string;
  }>;
}
```
