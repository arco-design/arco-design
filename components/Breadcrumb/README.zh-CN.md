`````
组件 / 导航

# 面包屑 Breadcrumb

面包屑是辅助导航模式，用于识别页面在层次结构内的位置，并根据需要向上返回。
`````

%%Content%%

## API

### Breadcrumb

|参数名|描述|类型|默认值|
|---|---|---|---|
|maxCount|最多渲染的面包屑数量|number |`-`|
|separator|指定分割符|string \| ReactNode |`<IconObliqueLine />`|
|className|节点类名|string \| string[] |`-`|
|routes|直接设置下拉菜单|[RouteProps](#routeprops)[] |`-`|
|style|节点样式|CSSProperties |`-`|
|itemRender|routes 时生效，自定义渲染面包屑|(route: [RouteProps](#routeprops), routes: [RouteProps](#routeprops)[], paths: string[]) => ReactNode |`-`|

### Breadcrumb.Item

|参数名|描述|类型|默认值|
|---|---|---|---|
|droplist|下拉菜单的内容，等同于下拉菜单组件的 droplist 属性|[DropdownProps](dropdown#dropdown)['droplist'] |`-`|
|className|节点类名|string \| string[] |`-`|
|dropdownProps|下拉菜单的属性 [DropdownProps](/react/components/dropdown)|[DropdownProps](dropdown#dropdown) |`-`|
|style|节点样式|CSSProperties |`-`|

### RouteProps

```js
export interface RouteProps {
  path: string;
  breadcrumbName: string;
  children?: Array<{
    path: string;
    breadcrumbName: string;
  }>;
}
```
