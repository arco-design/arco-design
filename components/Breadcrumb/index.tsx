import React, { ReactNode, useContext, PropsWithChildren, forwardRef } from 'react';
import cs from '../_util/classNames';
import Item from './item';
import { ConfigContext } from '../ConfigProvider';
import Menu from '../Menu';
import IconObliqueLine from '../../icon/react-icon/IconObliqueLine';
import omit from '../_util/omit';
import { RouteProps, BreadcrumbProps } from './interface';
import useMergeProps from '../_util/hooks/useMergeProps';

const defaultItemRender = (route: RouteProps, routes: RouteProps[], paths: string[]): ReactNode => {
  if (routes.indexOf(route) === routes.length - 1) {
    return <span>{route.breadcrumbName}</span>;
  }
  return <a href={`#/${paths.join('/').replace(/^\//, '')}`}>{route.breadcrumbName}</a>;
};

const defaultProps: BreadcrumbProps = {
  separator: <IconObliqueLine />,
};

function Breadcrumb(baseProps: PropsWithChildren<BreadcrumbProps>, ref) {
  const { getPrefixCls, componentConfig, rtl } = useContext(ConfigContext);
  const props = useMergeProps<PropsWithChildren<BreadcrumbProps>>(
    baseProps,
    defaultProps,
    componentConfig?.Breadcrumb
  );
  const { className, children, style, routes, maxCount, separator, ...rest } = props;

  const prefixCls = getPrefixCls('breadcrumb');
  const itemRender = 'itemRender' in props ? props.itemRender : defaultItemRender;

  const Ellipses = (
    <span aria-label="ellipses of breadcrumb items" className={`${prefixCls}-item-ellipses`}>
      ...
    </span>
  );
  const Separator = (
    <span aria-hidden className={`${prefixCls}-item-separator`}>
      {separator}
    </span>
  );

  const getValidChild = (itemToRender: ReactNode, delta: number, index: number) => {
    const SeparatorWithKey = React.cloneElement(Separator, { key: `${index}_separator` });

    // Show ellipses
    if (delta > 0) {
      if (index === 0) {
        return [itemToRender, SeparatorWithKey, Ellipses];
      }

      if (index > delta) {
        return [SeparatorWithKey, itemToRender];
      }

      return null;
    }

    return index === 0 ? [itemToRender] : [SeparatorWithKey, itemToRender];
  };

  const getItemsByRoute = () => {
    const paths = [];
    const delta = routes.length - maxCount;

    return routes.map((route, index) => {
      paths.push((route.path || '').replace(/^\//, ''));

      const droplist = route.children ? (
        <Menu>
          {route.children.map((item) => {
            return (
              <Menu.Item key={item.path || item.breadcrumbName}>
                {itemRender(item, routes, paths)}
              </Menu.Item>
            );
          })}
        </Menu>
      ) : null;

      return getValidChild(
        <Item prefixCls={prefixCls} key={route.path || route.breadcrumbName} droplist={droplist}>
          {itemRender(route, routes, paths)}
        </Item>,
        delta,
        index
      );
    });
  };

  const getItemsByChildren = () => {
    const delta = React.Children.toArray(children).length - maxCount;
    return React.Children.map(children, (child: React.ReactElement, index) => {
      return (
        child &&
        getValidChild(
          React.cloneElement(child, {
            prefixCls,
          }),
          delta,
          index
        )
      );
    });
  };

  return (
    <div
      role="list"
      ref={ref}
      style={style}
      className={cs(prefixCls, { [`${prefixCls}-rtl`]: rtl }, className)}
      {...omit(rest, ['itemRender'])}
    >
      {routes && routes.length ? getItemsByRoute() : getItemsByChildren()}
    </div>
  );
}

const ForwardRefBreadcrumb = forwardRef<unknown, BreadcrumbProps>(Breadcrumb);

const BreadcrumbComponent = ForwardRefBreadcrumb as typeof ForwardRefBreadcrumb & {
  Item: typeof Item;
};

BreadcrumbComponent.displayName = 'Breadcrumb';

BreadcrumbComponent.Item = Item;

export default BreadcrumbComponent;

export { BreadcrumbProps };
