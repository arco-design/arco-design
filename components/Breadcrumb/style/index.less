@import './token.less';

@breadcrumb-prefix-cls: ~'@{prefix}-breadcrumb';

.@{breadcrumb-prefix-cls} {
  display: inline-block;
  font-size: @breadcrumb-size-font-size;
  color: var(~'@{arco-cssvars-prefix}-color-text-2');

  &-icon {
    color: var(~'@{arco-cssvars-prefix}-color-text-2');
  }

  &-item {
    display: inline-flex;
    align-items: center;
    padding: 0 @breadcrumb-padding-text-horizontal;
    vertical-align: middle;
    line-height: @breadcrumb-size-text-height;
    color: @breadcrumb-color-text;

    > .@{prefix}-icon {
      color: @breadcrumb-color-icon;
    }

    a,
    &[href] {
      display: inline-block;
      border-radius: @breadcrumb-border-text-radius_hover;
      padding: 0 @breadcrumb-padding-text-horizontal;
      margin: 0 -@breadcrumb-padding-text-horizontal;
      text-decoration: none;
      color: @breadcrumb-color-link-text;
      background-color: @breadcrumb-color-bg;

      &:hover {
        background-color: @breadcrumb-color-bg_hover;
        color: @breadcrumb-color-link-text_hover;
      }

      &:focus-visible {
        box-shadow: 0 0 0 2px @breadcrumb-color-box-shadow;
      }
    }

    &:last-child {
      color: @breadcrumb-color-text_active;
      font-weight: @breadcrumb-font-weight_active;
    }

    &-ellipses {
      display: inline-block;
      position: relative;
      top: -3px;
      padding: 0 @breadcrumb-padding-text-horizontal;
      color: var(~'@{arco-cssvars-prefix}-color-text-2');
    }

    &-separator {
      display: inline-block;
      margin: 0 @breadcrumb-margin-separator-horizontal;
      vertical-align: middle;
      line-height: @breadcrumb-size-text-height;
      color: @breadcrumb-color-separator;
    }

    &-with-dropdown {
      cursor: pointer;
    }

    &-dropdown-icon {
      font-size: @breadcrumb-size-dropdown-icon;
      margin-left: @breadcrumb-margin-dropdown-icon-left;
      color: @breadcrumb-color-dropdown-icon;

      &-active svg {
        transform: rotate(180deg);
      }
    }
  }
}

.@{breadcrumb-prefix-cls}-rtl {
  .@{breadcrumb-prefix-cls}-item-dropdown-icon {
    margin-left: 0;
    margin-right: @breadcrumb-margin-dropdown-icon-left;
  }
}
