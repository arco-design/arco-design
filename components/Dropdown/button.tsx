import React, { ReactNode, ReactElement, forwardRef, useContext } from 'react';
import ArcoButton from '../Button';
import Dropdown from './index';
import IconMore from '../../icon/react-icon/IconMore';
import { DropdownButtonProps } from './interface';
import { ConfigContext } from '../ConfigProvider';
import useMergeProps from '../_util/hooks/useMergeProps';
import { pickDataAttributes } from '../_util/pick';

const defaultProps: DropdownButtonProps = {
  position: 'br',
  trigger: 'hover',
  type: 'default',
  icon: <IconMore />,
  unmountOnExit: true,
};

function Button(baseProps: DropdownButtonProps, ref) {
  const { componentConfig } = useContext(ConfigContext);
  const props = useMergeProps<DropdownButtonProps>(
    baseProps,
    defaultProps,
    componentConfig?.['Dropdown.Button']
  );
  const {
    className,
    style,
    children,
    disabled,
    droplist,
    position = 'br',
    trigger = 'hover',
    type = 'default',
    size,
    icon = <IconMore />,
    unmountOnExit = true,
    onClick,
    onVisibleChange,
    buttonProps,
    buttonsRender,
  } = props;

  let leftButton: ReactNode = (
    <ArcoButton disabled={disabled} type={type} size={size} onClick={onClick} {...buttonProps}>
      {children}
    </ArcoButton>
  );
  let rightButton: ReactNode = (
    <ArcoButton disabled={disabled} type={type} size={size} icon={icon} />
  );

  if (buttonsRender) {
    [leftButton, rightButton] = buttonsRender([leftButton, rightButton]);
  }

  const disableTrigger =
    disabled ||
    !rightButton ||
    ((rightButton as ReactElement).props && (rightButton as ReactElement).props.loading);

  return (
    <ArcoButton.Group className={className} style={style} ref={ref} {...pickDataAttributes(props)}>
      {leftButton}
      <Dropdown
        disabled={disabled}
        droplist={droplist}
        position={position}
        trigger={trigger}
        unmountOnExit={unmountOnExit}
        onVisibleChange={onVisibleChange}
        triggerProps={{
          disabled: disableTrigger,
        }}
      >
        {rightButton}
      </Dropdown>
    </ArcoButton.Group>
  );
}

const ButtonComponent = forwardRef(Button);

ButtonComponent.displayName = 'DropdownButton';

export default ButtonComponent;

export { DropdownButtonProps };
