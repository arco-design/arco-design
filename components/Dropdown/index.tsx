import React, { ReactElement, useContext, useRef } from 'react';
import Trigger, { EventsByTriggerNeed } from '../Trigger';
import Button from './button';
import { ConfigContext } from '../ConfigProvider';
import cs from '../_util/classNames';
import useMergeValue from '../_util/hooks/useMergeValue';
import omit from '../_util/omit';
import pick from '../_util/pick';
import { DropdownProps } from './interface';
import useMergeProps from '../_util/hooks/useMergeProps';

const defaultProps: DropdownProps = {
  position: 'bl',
  trigger: 'hover',
  unmountOnExit: true,
};

function Dropdown(baseProps: DropdownProps, _) {
  const { getPrefixCls, componentConfig } = useContext(ConfigContext);
  const props = useMergeProps<DropdownProps>(baseProps, defaultProps, componentConfig?.Button);
  const {
    trigger,
    droplist,
    children,
    position,
    disabled,
    unmountOnExit,
    triggerProps,
    getPopupContainer,
    onVisibleChange,
    ...rest
  } = props;

  const prefixCls = getPrefixCls('dropdown');

  const triggerRef = useRef(null);
  const [popupVisible, setPopupVisible] = useMergeValue(false, {
    defaultValue: props.defaultPopupVisible,
    value: props.popupVisible,
  });

  const getPopupContent = () => {
    return React.Children.only(droplist || <span />) as React.ReactElement;
  };

  const changePopupVisible = (visible: boolean) => {
    setPopupVisible(visible);
    onVisibleChange && onVisibleChange(visible);
    triggerProps && triggerProps.onVisibleChange && triggerProps.onVisibleChange(visible);
  };

  const handleVisibleChange = (visible: boolean) => {
    if (visible !== popupVisible) {
      changePopupVisible(visible);
    }
  };

  const renderPopup = () => {
    const content = getPopupContent();
    return content && content.props.isMenu
      ? React.cloneElement(content as ReactElement, {
          prefixCls: `${prefixCls}-menu`,
          inDropdown: true,
          selectable: false,
          onClickMenuItem: (key, event) => {
            let returnValueOfOnClickMenuItem = null;

            // Trigger onClickMenuItem first
            const content = getPopupContent();
            if (content.props.onClickMenuItem) {
              returnValueOfOnClickMenuItem = content.props.onClickMenuItem(key, event);
            }

            // Set focus to avoid onblur
            const child = triggerRef.current && triggerRef.current.getRootElement();
            child && child.focus && child.focus();

            // Trigger onVisibleChange. Outer component can determine whether to change the state based on the current visibility value.
            if (returnValueOfOnClickMenuItem instanceof Promise) {
              returnValueOfOnClickMenuItem.finally(() => changePopupVisible(false));
            } else if (returnValueOfOnClickMenuItem !== false) {
              changePopupVisible(false);
            }
          },
        })
      : content;
  };

  return (
    <Trigger
      ref={(ref) => (triggerRef.current = ref)}
      classNames="slideDynamicOrigin"
      childrenPrefix={prefixCls}
      trigger={trigger}
      popup={renderPopup}
      mouseEnterDelay={400}
      mouseLeaveDelay={400}
      disabled={disabled}
      unmountOnExit={unmountOnExit}
      position={position}
      popupVisible={popupVisible}
      popupAlign={{
        left: 4,
        right: 4,
        top: 4,
        bottom: 4,
      }}
      getPopupContainer={getPopupContainer}
      alignPoint={trigger === 'contextMenu'}
      {...pick(rest, EventsByTriggerNeed)}
      {...omit(triggerProps, ['onVisibleChange'])}
      onVisibleChange={handleVisibleChange}
    >
      {React.isValidElement(children)
        ? React.cloneElement(children, {
            disabled,
            className: cs(
              {
                [`${prefixCls}-popup-visible`]: popupVisible,
              },
              children.props.className
            ),
          })
        : children}
    </Trigger>
  );
}

const ForwardRefDropdown: React.ForwardRefExoticComponent<
  DropdownProps & React.RefAttributes<unknown>
> = React.forwardRef<unknown, DropdownProps>(Dropdown);

const DropdownComponent = ForwardRefDropdown as typeof ForwardRefDropdown & {
  Button: typeof Button;
};

DropdownComponent.displayName = 'Dropdown';

DropdownComponent.Button = Button;

export default DropdownComponent;

export { DropdownProps };
