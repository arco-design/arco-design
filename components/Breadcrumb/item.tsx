import React, { PropsWithChildren, useState } from 'react';
import { pickDataAttributes } from '../_util/pick';
import cs from '../_util/classNames';
import Dropdown from '../Dropdown';
import IconDown from '../../icon/react-icon/IconDown';
import omit from '../_util/omit';
import { BreadCrumbItemProps } from './interface';
import { isString } from '../_util/is';

function Item(props: PropsWithChildren<BreadCrumbItemProps>) {
  const {
    children,
    style,
    className,
    prefixCls,
    droplist,
    dropdownProps,
    href,
    onClick,
    tagName = 'div',
    ...rest
  } = props;
  const [dropdownVisible, setDropdownVisible] = useState(false);

  const TagName = isString(href) ? 'a' : tagName;

  const dom = (
    <TagName
      href={href}
      onClick={onClick}
      role="listitem"
      style={style}
      className={cs(
        `${prefixCls}-item`,
        {
          [`${prefixCls}-item-with-dropdown`]: droplist,
        },
        className
      )}
      {...pickDataAttributes(rest)}
    >
      {children}
      {droplist && (
        <span
          aria-hidden
          className={cs(`${prefixCls}-item-dropdown-icon`, {
            [`${prefixCls}-item-dropdown-icon-active`]: dropdownVisible,
          })}
        >
          <IconDown />
        </span>
      )}
    </TagName>
  );

  return droplist ? (
    <Dropdown
      droplist={droplist}
      onVisibleChange={(visible) => {
        setDropdownVisible(visible);
        dropdownProps && dropdownProps.onVisibleChange && dropdownProps.onVisibleChange(visible);
      }}
      {...omit(dropdownProps, ['onVisibleChange'])}
    >
      {dom}
    </Dropdown>
  ) : (
    dom
  );
}

Item.displayName = 'BreadcrumbItem';

export default Item;

export { BreadCrumbItemProps };
