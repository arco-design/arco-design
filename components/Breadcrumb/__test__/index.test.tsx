import React from 'react';
import { mount } from 'enzyme';
import mountTest from '../../../tests/mountTest';
import componentConfigTest from '../../../tests/componentConfigTest';
import Breadcrumb from '..';
import { DropdownProps } from '../../Dropdown';

const BreadcrumbItem = Breadcrumb.Item;

mountTest(Breadcrumb);
componentConfigTest(Breadcrumb, 'Breadcrumb');

describe('Breadcrumb', () => {
  it('breadcrumb separator correctly', () => {
    const wrapper = mount(
      <Breadcrumb separator="=>">
        <BreadcrumbItem>Home</BreadcrumbItem>
        <BreadcrumbItem>Article</BreadcrumbItem>
        <BreadcrumbItem>Technology</BreadcrumbItem>
      </Breadcrumb>
    );
    expect(wrapper.find('.arco-breadcrumb-item-separator').at(0).text()).toBe('=>');
  });

  it('breadcrumb render empty children correctly', () => {
    const wrapper = mount(<Breadcrumb />);
    expect(wrapper.find('.arco-breadcrumb')).toHaveLength(1);
    expect(wrapper.find('.arco-breadcrumb-item')).toHaveLength(0);
  });

  it('supprt droplist correctly', () => {
    const arr = new Array(3).fill(1);
    const dropdownProps: DropdownProps = { trigger: 'click', position: 'bl' };
    const wrapper = mount(
      <Breadcrumb>
        {arr.map((_, index) => {
          if (index === 1) {
            return (
              <BreadcrumbItem
                droplist={<div>droplist-{index}</div>}
                key={index}
                dropdownProps={dropdownProps}
              >
                {index}
              </BreadcrumbItem>
            );
          }
          return <BreadcrumbItem key={index}>{index}</BreadcrumbItem>;
        })}
      </Breadcrumb>
    );
    expect(wrapper.find('IconDown')).toHaveLength(1);

    const DropdownElemProps = wrapper.find('Dropdown').props();
    Object.keys(dropdownProps).forEach((key) => {
      expect(DropdownElemProps[key]).toEqual(dropdownProps[key]);
    });
  });

  it('support routes correctly', () => {
    const routes = new Array(5).fill(1).map((_, index) => {
      if (index % 2 === 0) {
        return {
          path: `/${index}`,
          breadcrumbName: `name-${index}`,
          children: new Array(3).fill(2).map((_, index) => ({
            path: `/${index}`,
            breadcrumbName: `children-${index}`,
          })),
        };
      }
      return {
        path: `/${index}`,
        breadcrumbName: `name-${index}`,
      };
    });

    const wrapper = mount(<Breadcrumb routes={routes} />);
    expect(wrapper.find('Dropdown')).toHaveLength(3);
    expect(wrapper.find('.arco-breadcrumb-item a').at(2).text()).toEqual(`name-2`);

    expect(wrapper.find('.arco-breadcrumb-item a').at(2).prop('href')).toEqual('#/0/1/2');
  });

  it('support itemrender correctly', () => {
    const mockItemRender = jest.fn();
    const routes = new Array(5).fill(1).map((_, index) => {
      if (index % 2 === 0) {
        return {
          path: `/${index}`,
          breadcrumbName: `name-${index}`,
          children: new Array(3).fill(2).map((_, index) => ({
            path: `/${index}`,
            breadcrumbName: `children-${index}`,
          })),
        };
      }
      return {
        path: `/${index}`,
        breadcrumbName: `name-${index}`,
      };
    });

    mount(<Breadcrumb routes={routes} itemRender={mockItemRender} />);
    const allLength = 5 + 3 * 3;
    expect(mockItemRender).toHaveBeenCalledTimes(allLength);
  });

  it('support maxCount correctly', () => {
    const routes = new Array(10).fill(1).map((_, index) => ({
      path: `/${index}`,
      breadcrumbName: `name-${index}`,
    }));

    let maxCount = 3;
    const wrapper = mount(<Breadcrumb maxCount={maxCount} routes={routes} />);
    expect(wrapper.find('.arco-breadcrumb-item')).toHaveLength(3);
    expect(wrapper.find('.arco-breadcrumb-item-ellipses')).toHaveLength(1);

    maxCount = 5;
    const delta = routes.length - maxCount;
    wrapper.setProps({ maxCount });
    expect(wrapper.find('.arco-breadcrumb-item')).toHaveLength(5);
    expect(wrapper.find('.arco-breadcrumb-item').at(3).text()).toEqual(`name-${delta + 3}`);
  });
});
