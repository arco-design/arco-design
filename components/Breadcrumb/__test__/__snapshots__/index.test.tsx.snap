// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.Breadcrumb default 1`] = `
<div
  class="arco-breadcrumb"
  role="list"
/>
`;

exports[`ConfigProvider componentConfig.Breadcrumb set className globally 1`] = `
<div
  class="arco-breadcrumb global-default-class"
  role="list"
/>
`;

exports[`ConfigProvider componentConfig.Breadcrumb set className globally and component set className itself. 1`] = `
<div
  class="arco-breadcrumb component-class"
  role="list"
/>
`;

exports[`ConfigProvider componentConfig.Breadcrumb set data-* property 1`] = `
<div
  class="arco-breadcrumb"
  data-test-name="Breadcrumb"
  role="list"
/>
`;
